File: DemoProject/WidgetDesigners/Skills/SkillsDesigner.cs
using System;
using System.Linq;
using System.Web.UI;
using Telerik.Sitefinity.Web.UI;
using Telerik.Sitefinity.Web.UI.ControlDesign;
using System.Collections.Generic;

namespace SitefinityWebApp.WidgetDesigners.Skills
{
    /// <summary>
    /// Represents a designer for the <typeparamref name="SitefinityWebApp.Mvc.Controllers.SkillsController"/> widget
    /// </summary>
    public class SkillsDesigner : ControlDesignerBase
    {
        #region Properties
        /// <summary>
        /// Obsolete. Use LayoutTemplatePath instead.
        /// </summary>
        protected override string LayoutTemplateName
        {
            get
            {
                return string.Empty;
            }
        }

        /// <summary>
        /// Gets the layout template's relative or virtual path.
        /// </summary>
        public override string LayoutTemplatePath
        {
            get
            {
                if (string.IsNullOrEmpty(base.LayoutTemplatePath))
                    return SkillsDesigner.layoutTemplatePath;
                return base.LayoutTemplatePath;
            }
            set
            {
                base.LayoutTemplatePath = value;
            }
        }

        protected override HtmlTextWriterTag TagKey
        {
            get
            {
                return HtmlTextWriterTag.Div;
            }
        }
        #endregion

        #region Control references
        /// <summary>
        /// Gets the control that is bound to the Skill_1 property
        /// </summary>
        protected virtual Control Skill_1
        {
            get
            {
                return this.Container.GetControl<Control>("Skill_1", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Percentage_1 property
        /// </summary>
        protected virtual Control Percentage_1
        {
            get
            {
                return this.Container.GetControl<Control>("Percentage_1", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Skill_2 property
        /// </summary>
        protected virtual Control Skill_2
        {
            get
            {
                return this.Container.GetControl<Control>("Skill_2", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Percentage_2 property
        /// </summary>
        protected virtual Control Percentage_2
        {
            get
            {
                return this.Container.GetControl<Control>("Percentage_2", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Skill_3 property
        /// </summary>
        protected virtual Control Skill_3
        {
            get
            {
                return this.Container.GetControl<Control>("Skill_3", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Percentage_3 property
        /// </summary>
        protected virtual Control Percentage_3
        {
            get
            {
                return this.Container.GetControl<Control>("Percentage_3", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Skill_4 property
        /// </summary>
        protected virtual Control Skill_4
        {
            get
            {
                return this.Container.GetControl<Control>("Skill_4", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Percentage_4 property
        /// </summary>
        protected virtual Control Percentage_4
        {
            get
            {
                return this.Container.GetControl<Control>("Percentage_4", true);
            }
        }

        #endregion

        #region Methods
        protected override void InitializeControls(Telerik.Sitefinity.Web.UI.GenericContainer container)
        {
            // Place your initialization logic here
        }
        #endregion

        #region IScriptControl implementation
        /// <summary>
        /// Gets a collection of script descriptors that represent ECMAScript (JavaScript) client components.
        /// </summary>
        public override System.Collections.Generic.IEnumerable<System.Web.UI.ScriptDescriptor> GetScriptDescriptors()
        {
            var scriptDescriptors = new List<ScriptDescriptor>(base.GetScriptDescriptors());
            var descriptor = (ScriptControlDescriptor)scriptDescriptors.Last();

            descriptor.AddElementProperty("skill_1", this.Skill_1.ClientID);
            descriptor.AddElementProperty("percentage_1", this.Percentage_1.ClientID);
            descriptor.AddElementProperty("skill_2", this.Skill_2.ClientID);
            descriptor.AddElementProperty("percentage_2", this.Percentage_2.ClientID);
            descriptor.AddElementProperty("skill_3", this.Skill_3.ClientID);
            descriptor.AddElementProperty("percentage_3", this.Percentage_3.ClientID);
            descriptor.AddElementProperty("skill_4", this.Skill_4.ClientID);
            descriptor.AddElementProperty("percentage_4", this.Percentage_4.ClientID);

            return scriptDescriptors;
        }

        /// <summary>
        /// Gets a collection of ScriptReference objects that define script resources that the control requires.
        /// </summary>
        public override System.Collections.Generic.IEnumerable<System.Web.UI.ScriptReference> GetScriptReferences()
        {
            var scripts = new List<ScriptReference>(base.GetScriptReferences());
            scripts.Add(new ScriptReference(SkillsDesigner.scriptReference));
            return scripts;
        }
        #endregion

        #region Private members & constants
        public static readonly string layoutTemplatePath = "~/WidgetDesigners/Skills/SkillsDesigner.ascx";
        public const string scriptReference = "~/WidgetDesigners/Skills/SkillsDesigner.js";
        #endregion
    }
}
 

File: DemoProject/WidgetDesigners/Skills/SkillsDesigner.js
Type.registerNamespace("SitefinityWebApp.WidgetDesigners.Skills");

SitefinityWebApp.WidgetDesigners.Skills.SkillsDesigner = function (element) {
    /* Initialize Skill_1 fields */
    this._skill_1 = null;
    
    /* Initialize Percentage_1 fields */
    this._percentage_1 = null;
    
    /* Initialize Skill_2 fields */
    this._skill_2 = null;
    
    /* Initialize Percentage_2 fields */
    this._percentage_2 = null;
    
    /* Initialize Skill_3 fields */
    this._skill_3 = null;
    
    /* Initialize Percentage_3 fields */
    this._percentage_3 = null;
    
    /* Initialize Skill_4 fields */
    this._skill_4 = null;
    
    /* Initialize Percentage_4 fields */
    this._percentage_4 = null;
    
    /* Calls the base constructor */
    SitefinityWebApp.WidgetDesigners.Skills.SkillsDesigner.initializeBase(this, [element]);
}

SitefinityWebApp.WidgetDesigners.Skills.SkillsDesigner.prototype = {
    /* --------------------------------- set up and tear down --------------------------------- */
    initialize: function () {
        /* Here you can attach to events or do other initialization */
        SitefinityWebApp.WidgetDesigners.Skills.SkillsDesigner.callBaseMethod(this, 'initialize');
    },
    dispose: function () {
        /* this is the place to unbind/dispose the event handlers created in the initialize method */
        SitefinityWebApp.WidgetDesigners.Skills.SkillsDesigner.callBaseMethod(this, 'dispose');
    },

    /* --------------------------------- public methods ---------------------------------- */

    findElement: function (id) {
        var result = jQuery(this.get_element()).find("#" + id).get(0);
        return result;
    },

    /* Called when the designer window gets opened and here is place to "bind" your designer to the control properties */
    refreshUI: function () {
        var controlData = this._propertyEditor.get_control().Settings; /* JavaScript clone of your control - all the control properties will be properties of the controlData too */

        /* RefreshUI Skill_1 */
        jQuery(this.get_skill_1()).val(controlData.Skill_1);

        /* RefreshUI Percentage_1 */
        jQuery(this.get_percentage_1()).val(controlData.Percentage_1);

        /* RefreshUI Skill_2 */
        jQuery(this.get_skill_2()).val(controlData.Skill_2);

        /* RefreshUI Percentage_2 */
        jQuery(this.get_percentage_2()).val(controlData.Percentage_2);

        /* RefreshUI Skill_3 */
        jQuery(this.get_skill_3()).val(controlData.Skill_3);

        /* RefreshUI Percentage_3 */
        jQuery(this.get_percentage_3()).val(controlData.Percentage_3);

        /* RefreshUI Skill_4 */
        jQuery(this.get_skill_4()).val(controlData.Skill_4);

        /* RefreshUI Percentage_4 */
        jQuery(this.get_percentage_4()).val(controlData.Percentage_4);
    },

    /* Called when the "Save" button is clicked. Here you can transfer the settings from the designer to the control */
    applyChanges: function () {
        var controlData = this._propertyEditor.get_control().Settings;

        /* ApplyChanges Skill_1 */
        controlData.Skill_1 = jQuery(this.get_skill_1()).val();

        /* ApplyChanges Percentage_1 */
        controlData.Percentage_1 = jQuery(this.get_percentage_1()).val();

        /* ApplyChanges Skill_2 */
        controlData.Skill_2 = jQuery(this.get_skill_2()).val();

        /* ApplyChanges Percentage_2 */
        controlData.Percentage_2 = jQuery(this.get_percentage_2()).val();

        /* ApplyChanges Skill_3 */
        controlData.Skill_3 = jQuery(this.get_skill_3()).val();

        /* ApplyChanges Percentage_3 */
        controlData.Percentage_3 = jQuery(this.get_percentage_3()).val();

        /* ApplyChanges Skill_4 */
        controlData.Skill_4 = jQuery(this.get_skill_4()).val();

        /* ApplyChanges Percentage_4 */
        controlData.Percentage_4 = jQuery(this.get_percentage_4()).val();
    },

    /* --------------------------------- event handlers ---------------------------------- */

    /* --------------------------------- private methods --------------------------------- */

    /* --------------------------------- properties -------------------------------------- */

    /* Skill_1 properties */
    get_skill_1: function () { return this._skill_1; }, 
    set_skill_1: function (value) { this._skill_1 = value; },

    /* Percentage_1 properties */
    get_percentage_1: function () { return this._percentage_1; }, 
    set_percentage_1: function (value) { this._percentage_1 = value; },

    /* Skill_2 properties */
    get_skill_2: function () { return this._skill_2; }, 
    set_skill_2: function (value) { this._skill_2 = value; },

    /* Percentage_2 properties */
    get_percentage_2: function () { return this._percentage_2; }, 
    set_percentage_2: function (value) { this._percentage_2 = value; },

    /* Skill_3 properties */
    get_skill_3: function () { return this._skill_3; }, 
    set_skill_3: function (value) { this._skill_3 = value; },

    /* Percentage_3 properties */
    get_percentage_3: function () { return this._percentage_3; }, 
    set_percentage_3: function (value) { this._percentage_3 = value; },

    /* Skill_4 properties */
    get_skill_4: function () { return this._skill_4; }, 
    set_skill_4: function (value) { this._skill_4 = value; },

    /* Percentage_4 properties */
    get_percentage_4: function () { return this._percentage_4; }, 
    set_percentage_4: function (value) { this._percentage_4 = value; }
}

SitefinityWebApp.WidgetDesigners.Skills.SkillsDesigner.registerClass('SitefinityWebApp.WidgetDesigners.Skills.SkillsDesigner', Telerik.Sitefinity.Web.UI.ControlDesign.ControlDesignerBase);
